File: Sequence/q100-q199/q100-109/q109/q109.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        def sortedListToBST_helper(left: ListNode, right: ListNode) -> TreeNode:
            if left == right:
                return None
            fast, slow = left, left
            while fast != right and fast.next != right:
                fast = fast.next.next
                slow = slow.next
            root = TreeNode(slow.val)
            root.left = sortedListToBST_helper(left, slow)
            root.right = sortedListToBST_helper(slow.next, right)
            return root
        return sortedListToBST_helper(head, None)

File: Sequence/q100-q199/q100-109/q109/q109.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        return sortedListToBST_helper(head, NULL);
    }
    
    TreeNode* sortedListToBST_helper(ListNode* left, ListNode* right) {
        if (left == right) {
            return NULL;
        }
        // find the middle of linked list, using fast & slow pointer
        ListNode* fast = left;
        ListNode* slow = left;
        while (fast != right && fast->next != right) {
            fast = fast->next->next;
            slow = slow->next;
        }
        // here slow is the middle of linked list
        TreeNode* root = new TreeNode(slow->val);
        root->left = sortedListToBST_helper(left, slow);
        root->right = sortedListToBST_helper(slow->next, right);
        return root;
    }
};
